themes/default/content/docs/esc/other-integrations/terraform.md

```diff
@@ -0,0 +1,150 @@
+
+```bash
+$ terraform init
+```
+
+## Create an ESC environment with Terraform variables
+
+ESC integrates with Terraform by exporting environment variables from an opened environment. These environment variables typically come in two flavors: those used by Terraform providers and those used to provide values for Terraform input variables. The former are specific to each Terraform provider, while the latter are of the form [`TF_VAR_<variable name>`](https://developer.hashicorp.com/terraform/language/values/variables#environment-variables).
+
+For this guide, we'll create an environment that supplies both temporary credentials for the Terraform AWS provider as well as values for our confguration's `region` and `instance_type` input variables:
+
+```yaml
+values:
+  # Fetch temporary AWS credentials via OIDC
+  aws:
+    login:
+      fn::open::aws-login:
+        oidc:
+          duration: 1h
+          roleArn: <your-oidc-iam-role-arn>
+          sessionName: pulumi-environments-session
+  environmentVariables:
+    # Export AWS Credentials
+    AWS_ACCESS_KEY_ID: ${aws.login.accessKeyId}
+    AWS_SECRET_ACCESS_KEY: ${aws.login.secretAccessKey}
+    AWS_SESSION_TOKEN: ${aws.login.sessionToken}
+
+    # Export Terraform variables
+    TF_VAR_region: us-west-2
+    TF_VAR_instance_type: t3.small
+```
+
+## Run Terraform
+
+Once you've created the example environment, you can now use the Terraform CLI with `esc run` to invoke Terraform with credentials and configuration:
+
+```bash
+$ esc run [your-env-name] -i -- terraform apply
+```
+
+This will open your ESC environment, export its environment variables, and then invoke `terraform apply` with those environment variables.
+
+## Run Terraform with `-var-file`
+
+While simple Terraform input variables work well with environment variables, it can be difficult to set variables with complex types (e.g. arrays, objects). If your Terraform input variables have complex types, you can use `-var-file` to supply their values.
+
+```yaml
+values:
+  # Fetch temporary AWS credentials via OIDC
+  aws:
+    login:
+      fn::open::aws-login:
+        oidc:
+          duration: 1h
+          roleArn: <your-oidc-iam-role-arn>
+          sessionName: pulumi-environments-session
+  files:
+    TFVARS:
+      fn::toJSON:
+        region: us-west-2
+        instance_type: t3.small
+  environmentVariables:
+    # Export AWS Credentials
+    AWS_ACCESS_KEY_ID: ${aws.login.accessKeyId}
+    AWS_SECRET_ACCESS_KEY: ${aws.login.secretAccessKey}
+    AWS_SESSION_TOKEN: ${aws.login.sessionToken}
+```
+
+With this environment, the `esc` CLI will write the value of the `TFVARS` property to a temporary file, then export the path to that temporary file in the `TFVARS` environment variable. You can then pass the path to this file to the Terraform CLI. This is slightly more involved because the Terraform CLI requires that the contents of the vars file is either valid HCL or JSON, and JSON files must have a `.json` extension:
```

Comment:
Should we open an issue to add something like `fn::tfvars` to handle this?

Reply:
I think `fn::toHCL` would be a bit more consistent. We can open an issue. Might want more formats in the fullness of time as well depending on needs (`fn::toYAML`, `fn::toTOML`, etc)

Reply:
(I did open this the other day as well: https://github.com/pulumi/esc/issues/301)

Reply:
(opened https://github.com/pulumi/esc/issues/303 as well)